lib/linked_list.rb

```diff
@@ -12,120 +12,346 @@ def initialize(value, next_node = nil)
+    if @head.next
+      check_against_max = @head.next
+    else
+      return current_max
     end
-
-    # method to return the min value in the linked list
-    # returns the data value and not the node
-    def find_min
-      raise NotImplementedError
+    
+    while check_against_max
+      if check_against_max.data > current_max
+        current_max = check_against_max.data
+        if check_against_max.next
+          check_against_max = check_against_max.next
+        else
+          return current_max
+        end
+      else
+        if check_against_max.next
+          check_against_max = check_against_max.next
+        else
+          return current_max
+        end
+      end
     end
-
-
-    # method that returns the length of the singly linked list
-    def length
-      raise NotImplementedError
+    return current_max    
+  end
+  
+  # method to return the min value in the linked list
+  # returns the data value and not the node
+  def find_min
+    # checks if current is nil
+    if @head
+      current_min = @head.data
+    else
+      return nil
     end
-
-    # method that returns the value at a given index in the linked list
-    # index count starts at 0
-    # returns nil if there are fewer nodes in the linked list than the index value
-    def get_at_index(index)
-      raise NotImplementedError
+    
+    # checks if next is nil
+    if @head.next
+      check_against_min = @head.next
+    else
+      return current_min
     end
-
-    # method to print all the values in the linked list
-    def visit
-      raise NotImplementedError
+    
+    while check_against_min
+      if check_against_min.data < current_min
+        current_min = check_against_min.data
+        if check_against_min.next
+          check_against_min = check_against_min.next
+        else
+          return current_min
+        end
+      else
+        if check_against_min.next
+          check_against_min = check_against_min.next
+        else
+          return current_min
+        end
+      end
     end
-
-    # method to delete the first node found with specified value
-    def delete(value)
-      raise NotImplementedError
+    return current_min
+  end
+  
+  
+  # method that returns the length of the singly linked list
+  def length
+    count = 0
+    if @head
+      count += 1
+      current = @head
+      while current.next
+        current = current.next
+        count += 1
+      end
     end
-
-    # method to reverse the singly linked list
-    # note: the nodes should be moved and not just the values in the nodes
-    def reverse
-      raise NotImplementedError
+    return count
+  end
+  
+  # method that returns the value at a given index in the linked list
+  # index count starts at 0
+  # returns nil if there are fewer nodes in the linked list than the index value
+  def get_at_index(index)
+    if @head
+      count = 0
+      current_node = @head
+      while count < index
+        current_node = current_node.next

```

Comment:
What if `length < index`?  

```suggestion
        return nil if current_node.nil?
        current_node = current_node.next
```